torchdata/nodes/map.py

```diff
@@ -0,0 +1,195 @@
+# pyre-unsafe
+import queue
+import threading
+from typing import Callable, Iterator, List, Literal, Optional, TypeVar, Union
+
+import torch.multiprocessing as mp
+
+from torchdata.nodes import BaseNode, T
+
+from torchdata.nodes.exception_wrapper import ExceptionWrapper, StartupExceptionWrapper
+
+from ._apply_udf import _apply_udf
+
+from ._populate_queue import _populate_queue
+
+
+X = TypeVar("X")
+
+
+class Mapper(BaseNode[T]):
+    def __init__(
+        self,
+        source: BaseNode[X],
+        map_fn: Callable[[X], T],
+    ):
+        self.source = source
+        self.map_fn = map_fn
+
+    def iterator(self) -> Iterator[T]:
+        for item in self.source:
+            yield self.map_fn(item)
+
+
+class ParallelMapper(BaseNode[T]):
+    def __init__(
+        self,
+        source: BaseNode[X],
+        map_fn: Callable[[X], T],
+        num_workers: int,
+        in_order: bool = True,
+        method: Literal["thread", "process"] = "thread",
+    ):
+        self.source = source
+        self.udf = map_fn
+        self.num_workers = num_workers
+        self.in_order = in_order
+
+        self.in_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.out_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.sem = threading.BoundedSemaphore(value=2 * num_workers)
```

Comment:
Confirming my understanding, here `2 * num_workers` is essentially deciding the input queue size? Is `2` here a hyper-parameter to be tuned?

Reply:
Yes that's right, the entire value can be tuned, but we don't expose this right now